### PR TITLE
Include umbraco-package.json manifests in telemetry data

### PR DESCRIPTION
**Pull request by ronaldbarendse**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
PR https://github.com/umbraco/Umbraco-CMS/pull/15744 removed the legacy package manifest parser, but also removed the code from `PackagingService` that returned the combined the package migrations _and manifests_. This data is used in the telemetry service to report the installed packages (if the consent level of the installation is set to Basic or Detailed and the package hasn't opted-out), which is mainly used in the Marketplace to sort the packages by most active installs.

Another important detail that was missing, is being able to specify the (NuGet) package ID in the manifest, so the Marketplace can do a 1:1 match when aggregating the data.

Besides re-introducing the package manifests as installed packages and adding the package ID, I've ensured the ID is also exposed in the Management API endpoints. E.g. the `umbraco/management/api/v1/manifest/manifest` endpoint now returns:
```json
[
  {
    "name": "@umbraco-cms/backoffice",
    "id": null,
    "version": "14.0.0-rc2",
    "extensions": []
  },
  {
    "name": "Test",
    "id": "Umbraco.Cms.Test",
    "version": "14.0.0",
    "extensions": []
  }
]
```

Testing can be done by adding a `umbraco-package.json` file to e.g. `wwwroot\App_Plugins\Test`:
```json
{
  "name": "Test",
  "id": "Umbraco.Cms.Test",
  "version": "14.0.0",
  "extensions": []
}
```

And then making sure this is included in `ReportSiteJob.RunJobAsync()` (update Delay to 0 and add a breakpoint to this method) and the package ID is exposed in the Management API (all Manifest endpoints).

### CODE REVIEW

**Comment by ronaldbarendse** [comment]
The backoffice UI will still need to be updated to show the package ID, as that's currently missing:

Installed packages v14

Comparing it to v13, the package name is more visible/prominent and the version has a label, so maybe the styling could be slightly updated as well:

Installed packages v13

**Comment by Zeegaan** [comment]
Heyo 👋 
This won't make it in 14.0.0 sadly, so you'll need to fix the breaking changes in `IPackagingService`, should be fairly simple using Default implementation 😁

File: src/Umbraco.Infrastructure/Services/Implement/PackagingService.cs [hunk 1]
```diff
@@ -109,10 +131,14 @@ public class PackagingService : IPackagingService
     [Obsolete("Use DeleteCreatedPackageAsync instead. Scheduled for removal in Umbraco 15.")]
     public void DeleteCreatedPackage(int id, int userId = Constants.Security.SuperUserId)
     {
-        using ICoreScope scope = _coreScopeProvider.CreateCoreScope();
```
Comment (ronaldbarendse) [inline src/Umbraco.Infrastructure/Services/Implement/PackagingService.cs:112]: This scope was disposed without being completed, so the `CreatedPackageSchemaRepository` will delete the file, but rollback the database changes!

File: src/Umbraco.Infrastructure/BackgroundJobs/Jobs/ReportSiteJob.cs [hunk 2]
```diff
@@ -19,25 +20,37 @@ public class ReportSiteJob : IRecurringBackgroundJob
...
+    public ReportSiteJob(
+        ILogger<ReportSiteJob> logger,
+        ITelemetryService telemetryService,
+        IJsonSerializer jsonSerializer,
+        IHttpClientFactory httpClientFactory)
     {
         _logger = logger;
         _telemetryService = telemetryService;
         _jsonSerializer = jsonSerializer;
-        _httpClient = new HttpClient();
```
Comment (ronaldbarendse) [inline src/Umbraco.Infrastructure/BackgroundJobs/Jobs/ReportSiteJob.cs:40]: Using a static `HttpClient` without setting the `PooledConnectionLifetime` can cause issues with DNS changes: https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines. Using `IHttpClientFactory` instead will make sure connection pooling and lifetime management are done automatically 👍🏻

File: src/Umbraco.Infrastructure/Services/Implement/PackagingService.cs [hunk 3]
```diff
@@ -155,7 +183,8 @@ public async Task<PagedModel<PackageDefinition>> GetCreatedPackagesAsync(int ski
         using ICoreScope scope = _coreScopeProvider.CreateCoreScope(autoComplete: true);
         PackageDefinition[] packages = _createdPackages.GetAll().WhereNotNull().ToArray();
         var pagedModel = new PagedModel<PackageDefinition>(packages.Length, packages.Skip(skip).Take(take));
-        return await Task.FromResult(pagedModel);
```
Comment (ronaldbarendse) [inline src/Umbraco.Infrastructure/Services/Implement/PackagingService.cs:158]: I noticed we have quite a few occurrences of `await Task.FromResult()` spread throughout the codebase, which seems wasteful, as it wraps the result and adds an additional await (with all the extra async state machine code that's generated by the compiler)...

**Comment by ronaldbarendse** [comment]
> Heyo 👋 This won't make it in 14.0.0 sadly, so you'll need to fix the breaking changes in `IPackagingService`, should be fairly simple using Default implementation 😁

I've avoided any breaking changes now and added comments to highlight some additional bugs/issues that are fixed in this PR. Would be great if we can ship this in a patch, instead of waiting for the next minor 😇

**Comment by ronaldbarendse** [comment]
To help testing out the changes, you can put this composer/component in your project and check out the logged messages:
```c#
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.Telemetry;

internal sealed class LogPackageTelemetryComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.AddComponent<LogPackageTelemetryComponent>();
    }

    private sealed class LogPackageTelemetryComponent : IComponent
    {
        private readonly ITelemetryService _telemetryService;
        private readonly ILogger<LogPackageTelemetryComponent> _logger;

        public LogPackageTelemetryComponent(ITelemetryService telemetryService, ILogger<LogPackageTelemetryComponent> logger)
        {
            _telemetryService = telemetryService;
            _logger = logger;
        }

        public void Initialize()
        {
            var telemetryReportData = _telemetryService.GetTelemetryReportDataAsync().GetAwaiter().GetResult();
            if (telemetryReportData?.Packages is not null)
            {
                foreach (var packageTelemetry in telemetryReportData.Packages)
                {
                    _logger.LogInformation(
                        "Package telemetry for {PackageName} ({PackageId}) {PackageVersion}",
                        packageTelemetry.Name,
                        packageTelemetry.Id,
                        packageTelemetry.Version);
                }
            }
        }

        public void Terminate()
        { }
    }
}
```

On a side note: this does highlight the fact we don't have an `IAsyncCompontent` 😇

**Comment by Zeegaan** [comment]
Looks good, tests good 🚀

**Pull request merged**